crates/shared/src/trade_finding.rs

```diff
@@ -0,0 +1,37 @@
+//! A module for abstracting a component that can produce a quote with calldata
+//! for a specified token pair and amount.
+
+pub mod zeroex;
+
+use crate::price_estimation::Query;
+use ethcontract::U256;
+use thiserror::Error;
+
+/// Find a trade for a token pair.
+///
+/// This is similar to the `PriceEstimating` interface, but it expects calldata
+/// to also be produced.
+#[mockall::automock]
+#[async_trait::async_trait]
+pub trait TradeFinding: Send + Sync + 'static {
+    async fn get_trade(&self, query: &Query) -> Result<Trade, TradeError>;
+}
+
+/// A trade.
+pub struct Trade {
+    pub out_amount: U256,
+    pub gas_estimate: u64,
+    pub data: Vec<u8>,
```

Comment:
In this file it isn't super clear what `data` is. If it is an on chain call then doesn't it also need a target address and value (ether)?

Reply:
It does. It also needs an "allowance target" as well. I found that out when I tried to use it later 🙈

Reply:
Fixed - added remaining fields that are needed.